Fix wayland client interaction

TheNeikos opened the pull request.

TheNeikos committed:
Added ws_wayland_buffer_release

TheNeikos committed:
Added frame callback implementation to support changing buffers

TheNeikos committed:
Added null check against surface from wayland

matthiasbeyer commented on src/compositor/wayland/buffer.c:
```diff
@@ -334,3 +334,11 @@ end_access(
+ws_wayland_buffer_release(
+    struct ws_wayland_buffer* self
+) {
+    struct wl_resource* res = ws_wayland_obj_get_wl_resource(&self->wl_obj);
```
Could this possibly be `NULL`?

TheNeikos replied:
Yes, but what state should the application go if the buffer is NULL? There is no way to recover from that.

matthiasbeyer commented on src/compositor/wayland/surface.c:
```diff
@@ -337,6 +338,14 @@ surface_frame_cb(
+
+    struct ws_surface* surface = ws_surface_from_resource(resource);
+    surface->frame_callback = ws_wayland_client_create_resource(client,
+            &wl_callback_interface, 1, callback);
```
Please add more indentation here.

TheNeikos committed:
Added modifiers to keyboard

TheNeikos committed:
Removed the + 8 modifier from keycodes

This is something the client should do it seems. `weston-terminal` at least
seems to expect non offsetted key events... For the moment we have no choice but
to trust their code.

TheNeikos committed:
Removed double keyboard handle

neithernut commented:
I'll assume rhis is not WIP any more.

neithernut commented on src/compositor/keyboard.c:
```diff
@@ -239,6 +239,31 @@ ws_keyboard_send_key(
+    // ws_log(&log_ctx, LOG_DEBUG, "Depressed : %x", depressed);
+    // ws_log(&log_ctx, LOG_DEBUG, "Latched : %x", latched);
+    // ws_log(&log_ctx, LOG_DEBUG, "Locked : %x", locked);
+    // ws_log(&log_ctx, LOG_DEBUG, "Group : %x", group);
```
Those should either be removed or made real calls (instead of comments).

TheNeikos replied:
You don't want them to be real calls, so I will remove them.

matthiasbeyer replied:
Maybe you can change them to `ws_dbg()` calls? Then, they will only be included in a debug build...

neithernut replied:
A removal would be preferrable.

TheNeikos commented:
Okay, I fixed all the comments

matthiasbeyer commented on src/compositor/keyboard.c:
```diff
@@ -239,6 +239,26 @@ ws_keyboard_send_key(
     }
 
+    enum xkb_key_direction direction;
+    direction = state == WL_KEYBOARD_KEY_STATE_PRESSED ? XKB_KEY_DOWN : XKB_KEY_UP;
```
Uuhh... I'm not sure about the binding here. Can you please add parentheses around the condition?

matthiasbeyer commented on src/compositor/keyboard.c:
```diff
@@ -239,6 +239,26 @@ ws_keyboard_send_key(
+    locked = xkb_state_serialize_mods(self->xkb->state, XKB_STATE_LOCKED);
+    uint32_t group;
+    group = xkb_state_serialize_layout(self->xkb->state,
+                                        XKB_STATE_LAYOUT_EFFECTIVE);
```
Can you please move the variable declarations up and the variable settings down and align them? Like so:

``` c
uint32_t ae;
uint32_t b;

ae = foo();
b  = bar();
```

matthiasbeyer replied:
(Would result in more readable code ... at least for me)

matthiasbeyer commented on src/compositor/wayland/buffer.c:
```diff
@@ -120,6 +121,8 @@ end_access(
  *
  */
 
+static struct ws_logger_context log_ctx = { .prefix = "[Compositor/Surface]" };
```
Missing whitespace after `]` in string!

TheNeikos commented:
This has been changed again

neithernut commented:
ACK.
